File: 2016/Day13/main.py
from typing import Optional
import collections


FAVORITE_NUMBER = 1364

def is_wall(x: int, y: int) -> bool:
  if x < 0 or y < 0:
    return True
  formula = x*x + 3*x + 2*x*y + y + y*y + FAVORITE_NUMBER
  bit_count = bin(formula).count('1')
  return bit_count % 2 != 0

def bfs(start_x: int, start_y: int, target_x: int, target_y: int, max_steps: Optional[int] = None) -> int:
  queue = collections.deque([(start_x, start_y, 0)])
  visited = {(start_x, start_y)}

  while queue:
    x, y, steps = queue.popleft()

    if max_steps is None and x == target_x and y == target_y:
      return steps

    if max_steps is not None and steps >= max_steps:
      continue

    directions = [(1, 0), (-1, 0), (0, 1), (0, -1)]

    for dx, dy in directions:
      new_x, new_y = x + dx, y + dy

      if new_x >= 0 and new_y >= 0 and not is_wall(new_x, new_y) and (new_x, new_y) not in visited:
        visited.add((new_x, new_y))
        queue.append((new_x, new_y, steps + 1))

  if max_steps is not None:
    return len(visited)

  return -1

def main() -> None:
  start_x, start_y = 1, 1
  target_x, target_y = 31, 39

  part1 = bfs(start_x, start_y, target_x, target_y)
  print(f"Part 1: {part1}")

  part2 = bfs(start_x, start_y, target_x, target_y, max_steps=50)
  print(f"Part 2: {part2}")

if __name__ == "__main__":
  main()

File: 2016/Day13/go/main.go
package main

import (
	"fmt"
	"math/bits"
)

const favoriteNumber = 1364

func isWall(x, y int) bool {
	if x < 0 || y < 0 {
		return true
	}
	formula := x*x + 3*x + 2*x*y + y + y*y + favoriteNumber
	bitCount := bits.OnesCount(uint(formula))
	return bitCount%2 != 0
}

type Point struct {
	x, y int
}

type State struct {
	pt    Point
	steps int
}

func bfs(startX, startY, targetX, targetY, maxSteps int) int {
	queue := []State{}
	visited := make(map[Point]struct{})

	startPoint := Point{startX, startY}
	if isWall(startX, startY) {
		return -1
	}

	queue = append(queue, State{pt: startPoint, steps: 0})
	visited[startPoint] = struct{}{}

	directions := []Point{
		{1, 0},
		{-1, 0},
		{0, 1},
		{0, -1},
	}

	head := 0
	for head < len(queue) {
		current := queue[head]
		head++

		if maxSteps != 0 && current.steps >= maxSteps {
			continue
		}

		if maxSteps == 0 && current.pt.x == targetX && current.pt.y == targetY {
			return current.steps
		}

		for _, dir := range directions {
			nextX := current.pt.x + dir.x
			nextY := current.pt.y + dir.y
			nextPoint := Point{nextX, nextY}

			if nextX >= 0 && nextY >= 0 && !isWall(nextX, nextY) {
				if _, found := visited[nextPoint]; !found {
					visited[nextPoint] = struct{}{}
					queue = append(queue, State{pt: nextPoint, steps: current.steps + 1})
				}
			}
		}
	}

	if maxSteps != 0 {
		return len(visited)
	}

	return -1
}

func main() {
	startX, startY := 1, 1
	targetX, targetY := 31, 39

	fmt.Println("Part 1:", bfs(startX, startY, targetX, targetY, 0))
	fmt.Println("Part 2:", bfs(startX, startY, targetX, targetY, 50))
}
